### PR TITLE
Add tasks for building for musl LibC and all Dart-supported OSes

### PR DESCRIPTION
**Pull request by nex3**

### CODE REVIEW

**Comment by nex3** [comment]
@ntkme If you'd like to take a look at this as well, I'd appreciate it

File: lib/src/standalone/cli_platform.dart [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+            musl: musl)
+  };
+
+  /// The platform of the current Dart executable.
+  ///
+  /// This isn't necessarily the platform of the underlying OS, since 32-bit
+  /// executables can run on 64-bit operating systems and the choice of LibC is
+  /// OS-independent.
+  static final CliPlatform current = () {
+    // TODO(nweiz): Detect when we're running on musl LibC.
```
Comment (ntkme) [inline lib/src/standalone/cli_platform.dart:83]:
``` dart
import 'dart:io';
import 'package:path/path.dart' as p;
import 'package:native_stack_traces/elf.dart';

void main() {
  print(isMusl());
}

bool isMusl() {
  var elf = Elf.fromFile(Platform.resolvedExecutable);
  var section = elf?.namedSections('.interp').firstOrNull;
  if (section == null) return false;

  var file = File(Platform.resolvedExecutable).openSync();
  file.setPositionSync(section.offset);
  var interp = String.fromCharCodes(file.readSync(section.length - 1));
  file.closeSync();
  return p.basename(interp).startsWith('ld-musl-');
}
```

File: lib/src/standalone.dart [hunk 2]
```diff
@@ -266,50 +210,50 @@ Future<void> _buildPackage(String os, String arch) async {
...
+        "dartsdk-${platform.os}-${platform.arch}-release.tar.gz",
+    CliPlatform(
+      os: var os && (OperatingSystem.fuchsia || OperatingSystem.ios)
+    ) =>
+      fail(
+          "${os.toHumanString()} executables can only be generated when running "
+          "on ${os.toHumanString()}, because Dart doesn't distribute SDKs for "
+          "that platform."),
+    _ => "https://storage.googleapis.com/dart-archive/channels/"
+        "${isDevSdk ? "dev" : "stable"}/release/$dartVersion/sdk/"
```
Comment (ntkme) [inline lib/src/standalone.dart:242]: There is also `beta` channel.

**Review by ntkme — APPROVED** [review]
Looks good to me.

File: lib/src/standalone/cli_platform.dart [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+
+  CliPlatform(this.os, this.arch, {bool musl = false}) : isMusl = musl {
+    if (!_abiStrings.contains('${os}_$arch')) {
+      fail("Unknown or unsupported platform $os-$arch!");
+    }
+
+    if (musl && !os.isLinux) fail("musl LibC only supports Linux.");
+  }
+
+  int get hashCode => os.hashCode ^ arch.hashCode ^ isMusl.hashCode;
```
Comment (jathak) [inline lib/src/standalone/cli_platform.dart:129]: Is this still the preferred way to do this, or should this use `Object.hash`?

File: test/standalone_test.dart [hunk 4]
```diff
@@ -374,46 +374,85 @@ void main() {
...
+    group("iOS", () {
+      test("64-bit ARM", () async {
+        await (await grind(["pkg-standalone-ios-arm64"])).shouldExit(0);
+        await archive("ios", "arm64").validate();
+      });
+
+      test("64-bit x86", () async {
+        await (await grind(["pkg-standalone-ios-x64"])).shouldExit(0);
+        await archive("ios", "x64").validate();
+      });
```
Comment (jathak) [inline test/standalone_test.dart:455]: What are these builds useful for, if we can't build native binaries cross-platform and there's no iOS Dart SDK? Also, what x86 device runs iOS?

**Review by jathak — APPROVED** [review]
LGTM, though I'm a bit confused by the utility of the iOS stuff (unless I'm misunderstanding something about how they work)

File: test/standalone_test.dart (hunk 4 above)
Comment (nex3) [reply test/standalone_test.dart:455]: Good point. I'll just cut the iOS releases until that changes.

**Pull request merged**